sdk/python/kfp/dsl/component_factory.py

```diff
@@ -66,6 +66,13 @@ class ComponentInfo():
 
 def _python_function_name_to_component_name(name):
     name_with_spaces = re.sub(' +', ' ', name.replace('_', ' ')).strip(' ')
+    name_list = name_with_spaces.split(' ')
+
+    if name_list[-1].isdigit():
+        raise ValueError(
+            f'Invalid function name "{name}". The function name must not end in `_<int>`.'
```

Comment:
Can we focus on the backend change only without introduce a breaking change?

Reply:
Alternative design discussion continuing in [this](https://github.com/kubeflow/pipelines/pull/10979#discussion_r1697542558) thread.